File: packages/benchmarks/interrupt-bench/src/scenarios.ts
/**
 * Load and validate scenario JSON files from `scenarios/`.
 *
 * Lightweight runtime validation: confirms the required top-level keys are
 * present and types are roughly right. Heavier shape validation lives in the
 * `tests/scenarios.test.ts` vitest suite.
 */

import { readdirSync, readFileSync, type Stats, statSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import type { Scenario, ScenarioScriptStep } from "./types.ts";

const HERE = fileURLToPath(new URL(".", import.meta.url));
const SCENARIO_DIR = join(HERE, "..", "scenarios");
const EXPANSION_MULTIPLIER = 10;

type ScenarioVariant = {
  id: string;
  label: string;
  description: string;
  rewrite: (
    text: string,
    step: ScenarioScriptStep,
    index: number,
    scenario: Scenario,
  ) => string;
};

const EDGE_VARIANTS: ScenarioVariant[] = [
  {
    id: "polite",
    label: "polite framing",
    description: "Adds a polite setup to the first user turn.",
    rewrite: (text, _step, index) =>
      index === 0 ? `please help with this: ${text}` : text,
  },
  {
    id: "urgent",
    label: "urgent correction",
    description: "Frames the first turn as time-sensitive.",
    rewrite: (text, _step, index) =>
      index === 0 ? `this is time sensitive: ${text}` : text,
  },
  {
    id: "mobile",
    label: "mobile typing",
    description: "Adds a mobile-typed preface to the first turn.",
    rewrite: (text, _step, index) =>
      index === 0 ? `from my phone, quick note: ${text}` : text,
  },
  {
    id: "followup",
    label: "follow-up context",
    description: "Presents the first turn as continuing an earlier thread.",
    rewrite: (text, _step, index) =>
      index === 0 ? `following up from earlier: ${text}` : text,
  },
  {
    id: "quoted",
    label: "quoted handoff",
    description: "Quotes the first turn as a forwarded request.",
    rewrite: (text, _step, index) =>
      index === 0 ? `forwarded request:\n> ${text}` : text,
  },
  {
    id: "context",
    label: "extra context",
    description: "Adds operational context before the first turn.",
    rewrite: (text, _step, index, scenario) =>
      index === 0
        ? `context: ${scenario.category} interruption test\n${text}`
        : text,
  },
  {
    id: "ack",
    label: "ack requested",
    description: "Asks for a brief acknowledgement on the first turn.",
    rewrite: (text, _step, index) =>
      index === 0 ? `please keep the reply brief: ${text}` : text,
  },
  {
    id: "noisy",
    label: "chat noise",
    description: "Adds natural chat filler to the first turn.",
    rewrite: (text, _step, index) =>
      index === 0 ? `hey, sorry for the messy message, ${text}` : text,
  },
  {
    id: "boundary",
    label: "instruction boundary",
    description: "Marks the first turn as the beginning of user intent.",
    rewrite: (text, _step, index) =>
      index === 0 ? `user intent starts here:\n${text}` : text,
  },
  {
    id: "handoff",
    label: "teammate handoff",
    description: "Frames the first turn as delegated by another person.",
    rewrite: (text, _step, index) =>
      index === 0 ? `my teammate asked me to handle this: ${text}` : text,
  },
];

if (EDGE_VARIANTS.length !== EXPANSION_MULTIPLIER) {
  throw new Error(
    `InterruptBench expansion requires exactly ${EXPANSION_MULTIPLIER} variants, found ${EDGE_VARIANTS.length}`,
  );
}

function isScenarioShape(obj: unknown): obj is Scenario {
  if (!obj || typeof obj !== "object") return false;
  const o = obj as Record<string, unknown>;
  if (typeof o.id !== "string") return false;
  if (typeof o.category !== "string") return false;
  if (typeof o.interruptionType !== "string") return false;
  if (typeof o.weight !== "number") return false;
  if (
    !o.setup ||
    !o.script ||
    !o.expectedFinalState ||
    !o.expectedTrace ||
    !o.responseRubric
  )
    return false;
  return true;
}

function loadBaseScenarios(): Scenario[] {
  const out: Scenario[] = [];
  for (const category of readdirSync(SCENARIO_DIR)) {
    const catPath = join(SCENARIO_DIR, category);
    let stat: Stats;
    try {
      stat = statSync(catPath);
    } catch {
      continue;
    }
    if (!stat.isDirectory()) continue;
    for (const file of readdirSync(catPath)) {
      if (!file.endsWith(".json")) continue;
      const raw = readFileSync(join(catPath, file), "utf8");
      const parsed = JSON.parse(raw) as unknown;
      if (!isScenarioShape(parsed)) {
        throw new Error(`Scenario ${category}/${file} is malformed`);
      }
      out.push(parsed);
    }
  }
  out.sort((a, b) => a.id.localeCompare(b.id));
  return out;
}

function cloneScenario(scenario: Scenario): Scenario {
  return JSON.parse(JSON.stringify(scenario)) as Scenario;
}

function applyVariant(scenario: Scenario, variant: ScenarioVariant): Scenario {
  const expanded = cloneScenario(scenario);
  expanded.id = `${scenario.id}--edge-${variant.id}`;
  expanded.description = `${scenario.description ?? scenario.id} Edge variant: ${variant.description}`;
  expanded.script = scenario.script.map((step, index) => ({
    ...step,
    text: variant.rewrite(step.text, step, index, scenario),
  }));
  return expanded;
}

export function getBaseScenarioId(id: string): string {
  return id.replace(/--edge-[a-z-]+$/, "");
}

export function loadBaseScenarioSet(): Scenario[] {
  return loadBaseScenarios();
}

export function loadExpandedScenarios(): Scenario[] {
  const base = loadBaseScenarios();
  const expanded = base.flatMap((scenario) =>
    EDGE_VARIANTS.map((variant) => applyVariant(scenario, variant)),
  );
  if (expanded.length !== base.length * EXPANSION_MULTIPLIER) {
    throw new Error(
      `InterruptBench scenario expansion mismatch: expected ${base.length * EXPANSION_MULTIPLIER}, found ${expanded.length}`,
    );
  }
  return expanded;
}

export function loadScenarios(): Scenario[] {
  return [...loadBaseScenarios(), ...loadExpandedScenarios()];
}

export function loadScenarioById(id: string): Scenario | null {
  return loadScenarios().find((s) => s.id === id) ?? null;
}

export function countInterruptBenchScenarios(): {
  suite: "interrupt-bench";
  existing: number;
  added: number;
  total: number;
  multiplierAdded: number;
} {
  const base = loadBaseScenarios();
  const expanded = loadExpandedScenarios();
  return {
    suite: "interrupt-bench",
    existing: base.length,
    added: expanded.length,
    total: base.length + expanded.length,
    multiplierAdded: expanded.length / base.length,
  };
}

export function validateInterruptBenchScenarios(): {
  valid: boolean;
  total: number;
  uniqueIds: number;
  duplicateIds: string[];
  emptyScriptSteps: string[];
  expansionMatches: boolean;
} {
  const base = loadBaseScenarios();
  const expanded = loadExpandedScenarios();
  const all = [...base, ...expanded];
  const ids = new Set<string>();
  const duplicateIds = new Set<string>();
  const emptyScriptSteps: string[] = [];

  for (const scenario of all) {
    if (ids.has(scenario.id)) duplicateIds.add(scenario.id);
    ids.add(scenario.id);
    if (
      scenario.script.length === 0 ||
      scenario.script.some((step) => step.text.trim().length === 0)
    ) {
      emptyScriptSteps.push(scenario.id);
    }
  }

  const expansionMatches =
    expanded.length === base.length * EXPANSION_MULTIPLIER;

  return {
    valid:
      duplicateIds.size === 0 &&
      emptyScriptSteps.length === 0 &&
      expansionMatches,
    total: all.length,
    uniqueIds: ids.size,
    duplicateIds: [...duplicateIds],
    emptyScriptSteps,
    expansionMatches,
  };
}

File: packages/benchmarks/clawbench/clawbench/scenarios.py
"""Scenario loading and edge expansion helpers for ClawBench."""

from __future__ import annotations

import copy
import re
from pathlib import Path
from typing import Any

import yaml

CLAWBENCH_DIR = Path(__file__).resolve().parent.parent
SCENARIOS_DIR = CLAWBENCH_DIR / "scenarios"
EXPANSION_MULTIPLIER = 10

ScenarioDict = dict[str, Any]
Variant = tuple[str, str, str]

EDGE_VARIANTS: tuple[Variant, ...] = (
    ("polite", "Polite framing.", "Please help with this: {prompt}"),
    ("urgent", "Urgent operations framing.", "This is time sensitive: {prompt}"),
    ("mobile", "Mobile-message framing.", "Sent from mobile, quick note: {prompt}"),
    ("followup", "Follow-up thread framing.", "Following up from earlier: {prompt}"),
    ("quoted", "Forwarded quoted request.", "Forwarded request:\n> {prompt}"),
    ("context", "Extra operating context.", "Context: ClawBench edge case\n{prompt}"),
    ("brief", "Brevity preference.", "Keep this brief if you reply: {prompt}"),
    ("noisy", "Natural chat filler.", "Hey, sorry for the messy phrasing, {prompt}"),
    ("boundary", "Explicit user-intent boundary.", "User intent starts here:\n{prompt}"),
    ("handoff", "Teammate handoff framing.", "My teammate asked me to pass this along: {prompt}"),
)

if len(EDGE_VARIANTS) != EXPANSION_MULTIPLIER:
    raise RuntimeError(
        f"ClawBench expansion requires {EXPANSION_MULTIPLIER} variants, "
        f"found {len(EDGE_VARIANTS)}"
    )


def base_scenario_name(name: str) -> str:
    return re.sub(r"--edge-[a-z-]+$", "", name)


def _load_yaml(path: Path) -> ScenarioDict:
    with open(path, encoding="utf-8") as handle:
        data = yaml.safe_load(handle)
    if not isinstance(data, dict):
        raise ValueError(f"Scenario YAML at {path} did not parse to a dict")
    data.setdefault("name", path.stem)
    data["_base_name"] = data["name"]
    return data


def load_base_scenarios() -> list[ScenarioDict]:
    return [_load_yaml(path) for path in sorted(SCENARIOS_DIR.glob("*.y*ml"))]


def _apply_variant(scenario: ScenarioDict, variant: Variant) -> ScenarioDict:
    variant_id, description, template = variant
    out = copy.deepcopy(scenario)
    base_name = str(scenario.get("_base_name") or scenario.get("name") or "")
    out["name"] = f"{base_name}--edge-{variant_id}"
    out["_base_name"] = base_name
    out["description"] = (
        f"{str(scenario.get('description', '')).strip()} Edge variant: {description}"
    ).strip()
    out["prompt"] = template.format(prompt=str(scenario.get("prompt", "")))
    return out


def expand_scenarios(base_scenarios: list[ScenarioDict]) -> list[ScenarioDict]:
    expanded = [
        _apply_variant(scenario, variant)
        for scenario in base_scenarios
        for variant in EDGE_VARIANTS
    ]
    if len(expanded) != len(base_scenarios) * EXPANSION_MULTIPLIER:
        raise RuntimeError(
            "ClawBench scenario expansion mismatch: "
            f"expected {len(base_scenarios) * EXPANSION_MULTIPLIER}, found {len(expanded)}"
        )
    return expanded


def load_scenarios() -> list[ScenarioDict]:
    base = load_base_scenarios()
    return [*base, *expand_scenarios(base)]


def load_scenario(name_or_path: str) -> ScenarioDict:
    path = Path(name_or_path)
    if path.exists():
        return _load_yaml(path)

    for scenario in load_scenarios():
        if scenario.get("name") == name_or_path:
            return scenario

    available = sorted(str(scenario.get("name")) for scenario in load_scenarios())
    raise FileNotFoundError(f"Scenario '{name_or_path}' not found. Available: {available}")


def count_scenarios() -> dict[str, int | str | float]:
    base = load_base_scenarios()
    expanded = expand_scenarios(base)
    return {
        "suite": "clawbench",
        "existing": len(base),
        "added": len(expanded),
        "total": len(base) + len(expanded),
        "multiplierAdded": len(expanded) / len(base),
    }


def validate_scenarios() -> dict[str, object]:
    base = load_base_scenarios()
    expanded = expand_scenarios(base)
    all_scenarios = [*base, *expanded]
    ids: set[str] = set()
    duplicate_ids: set[str] = set()
    missing_prompt: list[str] = []
    missing_scoring: list[str] = []
    missing_tools: list[str] = []

    for scenario in all_scenarios:
        name = str(scenario.get("name", ""))
        if name in ids:
            duplicate_ids.add(name)
        ids.add(name)
        if not str(scenario.get("prompt", "")).strip():
            missing_prompt.append(name)
        if not scenario.get("scoring", {}).get("checks"):
            missing_scoring.append(name)
        if not scenario.get("tools"):
            missing_tools.append(name)

    expansion_matches = len(expanded) == len(base) * EXPANSION_MULTIPLIER
    return {
        "valid": not duplicate_ids
        and not missing_prompt
        and not missing_scoring
        and not missing_tools
        and expansion_matches,
        "total": len(all_scenarios),
        "uniqueIds": len(ids),
        "duplicateIds": sorted(duplicate_ids),
        "missingPrompt": missing_prompt,
        "missingScoring": missing_scoring,
        "missingTools": missing_tools,
        "expansionMatches": expansion_matches,
    }
